### PR TITLE
Attach data refactor and tuner bugs [4/n]

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Attach data refactor:
- Each of the trainer functions is now responsible for calling `attach_data` to register their `DataLoader`s/`DataModule`.
- No longer need to pass `DataLoaders` to `trainer._launch()`

Tuner bugs:
- `trainer.tuner.lr_find()` and `trainer.tuner.scale_batch_size()` now calls `trainer.tune()` internally. This solves a bug where the trainer state was not set properly as we set it in `trainer.tune()`
- `trainer.tune()` now has arguments for the different tuning functions.
- `trainer.tune()` now returns the tuning result. The result is a `{'scale_batch_size': Optional[int], 'lr_find': Optional[_LRFinder]}`

The tuning changes are tied to the `attach_data` changes because the tuner relied on the `setup_fit` function which has been removed in this PR.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/tuner/lr_finder.py [hunk 1]
```diff
@@ -366,6 +194,143 @@ def suggestion(self, skip_begin: int = 10, skip_end: int = 1):
             self._optimal_idx = None
 
 
+def lr_find(
```
Comment (carmocca) [inline pytorch_lightning/tuner/lr_finder.py:197]: Note to reviewers: This function was moved as we need the `_LRFinder` to be defined

**Comment by kaushikb11** [comment]
@carmocca 

Why is this one PR doing so many things?
```
Non-exhaustive typing
Moved _validate_data_hooks into ConfigValidator
Each of the trainer functions is now responsible of calling attach_data to register their DataLoaders/DataModule.
No longer need to pass DataLoaders to _launch
Deprecate trainer.tuner.lr_find() and trainer.tuner.scale_batch_size() in favor of trainer.tune(). This solves a bug where the trainer state was not set properly as we set it in trainer.tune()
trainer.tune() now has arguments for the different tuning functions.
trainer.tune() now returns the tuning result.
Fixed bug where scale_batch_size would fail if the number of trials was 0
```

**Comment by carmocca** [comment]
> Why is this one PR doing so many things?

All of those things are tied together. I just wrote the list of changes so it's easier for reviewers to follow.

If you have an idea of how it could be split, I don't mind splitting it further.

File: docs/source/advanced/training_tricks.rst [hunk 2]
```diff
@@ -112,18 +112,18 @@ search for batch sizes larger than the size of the training dataset.
     to `.fit()`.
 
 The scaling algorithm has a number of parameters that the user can control by
-invoking the trainer method `.scale_batch_size` themself (see description below).
+invoking passing the ``scale_batch_size`` argument:
```
Comment (SkafteNicki) [inline docs/source/advanced/training_tricks.rst:115]:
```suggestion
passing the ``scale_batch_size`` arguments to ``trainer.tune``:
```

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1095,16 +1101,39 @@ def tune(
...
+        # If you supply a datamodule you can't supply train_dataloader or val_dataloaders
+        if (train_dataloader is not None or val_dataloaders is not None) and datamodule is not None:
+            raise MisconfigurationException(
+                'You cannot pass `train_dataloader` or `val_dataloaders` to `trainer.tune(datamodule=...)`'
+            )
+
+        # links data to the trainer
+        self.data_connector.attach_data(
+            model, train_dataloader=train_dataloader, val_dataloaders=val_dataloaders, datamodule=datamodule
+        )
```
Comment (SkafteNicki) [inline pytorch_lightning/trainer/trainer.py:1126]: This seems redundant. Refactor into function?

File: pytorch_lightning/tuner/tuning.py [hunk 4]
```diff
@@ -33,44 +31,30 @@ def on_trainer_init(self, auto_lr_find: Union[str, bool], auto_scale_batch_size:
...
-        # setup data, etc...
-        self.trainer.train_loop.setup_fit(model, train_dataloader, val_dataloaders, datamodule)
-        # hook
-        self.trainer.data_connector.prepare_data(model)
+        model: 'pl.LightningModule',
+        scale_batch_size_kwargs: Optional[Dict[str, Any]] = None,
+        lr_find_kwargs: Optional[Dict[str, Any]] = None,
+    ) -> Dict[str, Optional[Union[int, _LRFinder]]]:
+        scale_batch_size_kwargs = scale_batch_size_kwargs or {}
+        lr_find_kwargs = lr_find_kwargs or {}
```
Comment (SkafteNicki) [inline pytorch_lightning/tuner/tuning.py:41]: this function is called as:
```python
result = self.tuner._tune(
            model, scale_batch_size_kwargs=scale_batch_size_kwargs or {}, lr_find_kwargs=lr_find_kwargs or {}
        )
```
the `kwargs or {}` therefore seems redundant.

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1126]: Why do you say it's redundant?

I don't think it's worth moving into a function as it is only called here. `trainer.fit` is also similar but the error message is different.

File: pytorch_lightning/tuner/tuning.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/tuner/tuning.py:41]: I'll remove it from the caller instead

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -1095,16 +1101,39 @@ def tune(
...
+
+        result = self.tuner._tune(
+            model, scale_batch_size_kwargs=scale_batch_size_kwargs or {}, lr_find_kwargs=lr_find_kwargs or {}
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:1129]:
```suggestion
            model, scale_batch_size_kwargs=scale_batch_size_kwargs, lr_find_kwargs=lr_find_kwargs
```

File: docs/source/advanced/training_tricks.rst [hunk 6]
```diff
@@ -112,18 +112,18 @@ search for batch sizes larger than the size of the training dataset.
...
 
     # Use default in trainer construction
-    trainer = Trainer()
-    tuner = Tuner(trainer)
+    trainer = Trainer(auto_scale_batch_size=True)
 
     # Invoke method
-    new_batch_size = tuner.scale_batch_size(model, *extra_parameters_here)
+    result = trainer.tune(model, scale_batch_size_kwargs={...})
+    new_batch_size = result['new_batch_size']
```
Comment (awaelchli) [inline docs/source/advanced/training_tricks.rst:124]: below you have it named "scale_batch_size" I think.

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 7]
```diff
@@ -79,24 +78,26 @@ def can_prepare_data(self):
...
+        self.attach_dataloaders(
+            model,
+            train_dataloader=train_dataloader,
+            val_dataloaders=val_dataloaders,
+            test_dataloaders=test_dataloaders,
+            predict_dataloaders=predict_dataloaders,
+        )
+        self.attach_datamodule(model, datamodule=datamodule)
+        # set local properties on the model
+        self.trainer.model_connector.copy_trainer_model_properties(model)
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/data_connector.py:100]: It was at the beginning of `_launch`. This does not work for `tune()` as we need these copied properties before we call `_launch`, that's why I moved it to here

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -908,10 +918,8 @@ def validate(
         model_provided = model is not None
         model = model or self.lightning_module
 
-        # Attach datamodule to get setup/prepare_data added to model before the call to it below
-        self.data_connector.attach_datamodule(model, datamodule)
-        #  Attach dataloaders (if given)
-        self.data_connector.attach_dataloaders(model, val_dataloaders=val_dataloaders)
+        # links data to the trainer
+        self.data_connector.attach_data(model, val_dataloaders=val_dataloaders, datamodule=datamodule)
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:922]: Good eye. I actually plan to open a PR adding this.

Not doing it yet as it would give myself conflicts

File: CHANGELOG.md [hunk 9]
```diff
@@ -206,6 +212,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated `@auto_move_data` in favor of `trainer.predict` ([#6993](https://github.com/PyTorchLightning/pytorch-lightning/pull/6993))
 
 
+- Deprecated `trainer.tuner.{lr_find,scale_batch_size}` in favor of `trainer.tune()` ([#7258](https://github.com/PyTorchLightning/pytorch-lightning/pull/7258))
+
+
```
[the remark this answers is not in the document]
Comment (carmocca) [reply CHANGELOG.md]: Un-deprecated it!

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/tuner/tuning.py [hunk 10]
```diff
@@ -119,47 +113,88 @@ def scale_batch_size(
...
+            val_dataloaders: Either a single Pytorch Dataloader or a list of them, specifying validation samples.
+                If the model has a predefined val_dataloaders method this will be skipped
+
+            datamodule: An instance of :class:`~pytorch_lightning.core.datamodule.LightningDataModule`.
+
+            min_lr: minimum learning rate to investigate
+
+            max_lr: maximum learning rate to investigate
+
+            num_training: number of learning rates to test
```
Comment (kaushikb11) [inline pytorch_lightning/tuner/tuning.py:166]: Not related to this PR. `num_training` param doesn't seem intuitive.

File: pytorch_lightning/trainer/trainer.py [hunk 11]
```diff
@@ -1094,17 +1100,38 @@ def tune(
             val_dataloaders: Either a single Pytorch Dataloader or a list of them, specifying validation samples.
                 If the model has a predefined val_dataloaders method this will be skipped
 
-            datamodule: A instance of :class:`LightningDataModule`.
+            datamodule: An instance of :class:`~pytorch_lightning.core.datamodule.LightningDataModule`.
+
+            scale_batch_size_kwargs: Arguments for :func:`~pytorch_lightning.tuner.lr_finder.lr_find`
+
+            lr_find_kwargs: Arguments for :func:`~pytorch_lightning.tuner.batch_size_scaling.scale_batch_size`
```
Comment (kaushikb11) [inline pytorch_lightning/trainer/trainer.py:1107]:
```suggestion
            scale_batch_size_kwargs: Arguments for :func:`~pytorch_lightning.tuner.batch_size_scaling.scale_batch_size`

            lr_find_kwargs: Arguments for :func:`~pytorch_lightning.tuner.lr_finder.lr_find`
```

File: tests/trainer/test_trainer.py [hunk 12]
```diff
@@ -1977,6 +1977,7 @@ def on_fit_start(self, trainer, pl_module: LightningModule) -> None:
 
 def test_exception_when_testing_or_validating_with_fast_dev_run(tmpdir):
     trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True)
+    trainer.fit(BoringModel())
```
Comment (kaushikb11) [inline tests/trainer/test_trainer.py:1980]:
```suggestion
	model = BoringModel()
    trainer.fit(model)
```
Nit

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: pytorch_lightning/tuner/tuning.py (hunk 10 above)
Comment (carmocca) [reply pytorch_lightning/tuner/tuning.py:166]: cc: @SkafteNicki 

I'm just moving the docstring around here

**Review by tchaton — APPROVED** [review]
LGMT !
